cpp/src/arrow/flight/server.h

```diff
@@ -45,9 +46,11 @@ class ARROW_EXPORT FlightDataStream {
  public:
   virtual ~FlightDataStream() = default;
 
-  // When the stream starts, send the schema.
   virtual std::shared_ptr<Schema> schema() = 0;
 
+  /// \brief Compute FlightPayload containing serialized RecordBatch schema
+  virtual Status GetSchemaPayload(FlightPayload* payload) = 0;
```

Comment:
(or is it @lihalite ?)

Reply:
Either will ping me!

I think in this case, it's OK, in that Flight explicitly states we transmit the schema first, then data; also, if we have a set of reasonable implementations of this interface, users should hopefully not feel a need to implement it themselves unless they did actually want control over the specifics.

Reply:
Right, but if you look at the RecordBatchStream implementation, it's doing non-trivial stuff with dictionaries.

Reply:
I think we'll have to develop some utility code to assist with these details, but it doesn't seem urgent for the moment. If you are creating a custom stream I am not sure right now how to protect the developer from the details of the stream protocol

Reply:
I did the necessary to hide the details in `IpcPayloadWriter`, and the converse is available for reading in `ipc::MessageReader` together with `RecordBatchStreamReader`. There shouldn't be a need to expose this at all.

Reply:
OK, how would you like to address this, sequence-wise -- is only the implementation here a problem, or is the public API also a problem? I think it would be much easier to fix both after getting this patch merged since we aren't releasing anytime soon

Reply:
Ok, let's fix this afterwards.